### PR TITLE
Skip redundant mapping by hash GC in Metadata.Builder

### PR DESCRIPTION
**Pull request by original-brownbear**
No need to check for mappings to GC if we didn't remove any use of any mapping.
Saves ~4% overall on the many-shards benchmarks initial indices bootstrap.

relates #77466

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java [hunk 1]
```diff
@@ -1307,9 +1316,27 @@ public Builder put(IndexMetadata indexMetadata, boolean incrementVersion) {
             if (unsetPreviousIndicesLookup(previous, indexMetadata)) {
                 previousIndicesLookup = null;
             }
+            maybeSetMappingPurgeFlag(previous, indexMetadata);
             return this;
         }
 
+        private void maybeSetMappingPurgeFlag(@Nullable IndexMetadata previous, IndexMetadata updated) {
+            if (previous == null) {
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:1324]: Can we shortcut here if `checkForUnusedMappings` is `true`?

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:1327]: ++ sure, 85183dd :)

**Review by martijnvg — APPROVED** [review]
Left one question, otherwise 👍

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java [hunk 2]
```diff
@@ -1247,6 +1247,13 @@ public static class Builder {
         private final ImmutableOpenMap.Builder<String, Custom> customs;
 
         private SortedMap<String, IndexAbstraction> previousIndicesLookup;
+
+        // If this is set to false we can skip checking #mappingsByHash for unused entries in #build(). Used as an optimization to save
+        // the rather expensive call to #purgeUnusedEntries when building from another instance and we know that no mappings can have
+        // become unused because no indices were updated or removed from this builder in a way that would cause unused entries in
+        // #mappingsByHash.
+        private boolean checkForUnusedMappings = true;
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:1255]: Then I have to put a true in the other constructor:

```java
private Builder(Map<String, MappingMetadata> mappingsByHash) {
```

-> I figured I'd be paranoid here and just do `true` by default and surgically insert `false` where I'm really sure :)

[the remark this answers is not in the document]
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:1255]: I see, that makes sense to me.

**Comment by original-brownbear** [comment]
Thanks Martijn!

**Pull request merged**